### ISSUE TITLE
Deprecate pipelined and move pipeline functionality into agentd

### ISSUE THREAD

**Issue by echlebek** [sensu/sensu-go#1353]
Like eventd and keepalived, pipelined represents a fixed size pipeline that we expect to handle any number of events. Unlike eventd and keepalived, pipelined has a somewhat lengthier request scope. Pipelined filters, handlers and mutators can result in processes being spawned, or requests being made to network services.

We may find it necessary to queue events going to pipelines, but each entity should not have its pipeline obstructed by other entities' filters, mutators or handlers.

Every session should include a pipeline goroutine that handles this. Investigate if there is a need for queueing, and if we need _persistent_ queueing.

**Comment by portertech** [comment]
Interesting, I find single queue + dispatcher compelling here for a number of reasons. Looking forward to discussing (but put our notes and findings here).

**Comment by portertech** [comment]
#1 feature of pipelined, no "fork bombing".

**Comment by portertech** [comment]
Making the fixed pipelined concurrency limit configurable would help.

**Comment by echlebek** [comment]
Sorry for taking so long to respond here.

This issue, and the ones similar to it, are driven by the goal of achieving as much parallelism in Sensu as possible. The best way to do that is to minimize information sharing; a shared-nothing system can achieve optimal parallelism.

In our current design, we have to process every event in pipelined. That means that every agent's session must synchronize with pipelined, regardless of the content of the event. However, there is no such requirement to do so; we do not need to serialize events globally.

Instead, events only need to be serialized on a per-entity basis. This means we can process events inside of our session goroutines, or in another separate goroutine for each session. As such, we can reduce blocking for events significantly in the common case.

Another requirement is that we do not exec `N` processes in the execution of filters, mutators and handlers. This was the original stated intent of pipelined, which it achieves; pipelined has a fixed number of goroutines and so can only exec as many processes as there are goroutines.

This requirement can be solved by using a worker pool or token bucket. We simply need to constrain the number of processes that are forked in executing filters, mutators and handlers.

To recap, we can limit information sharing to filters, mutators and handlers that need to be executed as external processes, etcd writes notwithstanding.
